apis/v1alpha2/httproute_types.go

```diff
@@ -406,15 +409,24 @@ type HTTPRouteMatch struct {
 	// +kubebuilder:default={type: "Prefix", value: "/"}
 	Path *HTTPPathMatch `json:"path,omitempty"`
 
-	// Headers specifies a HTTP request header matcher.
+	// Headers specifies HTTP request header matchers. Multiple match values are
+	// ANDed together, meaning, a request must match all the specified headers
+	// to select the route.
+	//
+	// If multiple entries specify equivalent header names, the first entry with
+	// an equivalent name MUST be given precedence. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered
+	// equivalent.
```

Comment:
Question for clarification:
```yaml
headers:
- name: foo
  value: bar
- name: Foo
   value: bar1
```

What happens if a request with header `foo: bar1` comes in? Is it matched or not?
I'm not sure if I clearly understand the meaning of "precedence" here.

Reply:
Yeah, this could be clarified better. I was trying to say `foo: bar` should be given precedence here since it appears first in the list. That means that `foo: bar1` would not match (unless type was "Prefix").

Reply:
This is in conflict with the general rule, which is "multiple match values are ANDed". So I'm not sure that we need to add the complexity of a special case for duplicates (as a quality of implementation issue, maybe validation could reject this case or a controller could emit a warning).

Reply:
so in the above example no requests should match because the result of an AND here would always be false?

Reply:
> so in the above example no requests should match because the result of an AND here would always be false?

Yup, exactly.

Reply:
That is hard to understand based on the current language. I've a suggestion to improve the language a little bit here.
I think having validation to prevent the user from providing two different values for the same header with type `Exact` is a good idea. We should open an issue to track it if you all

Reply:
So I spent awhile trying to find the right language here, and everything I typed out was either too obvious, too confusing, or both. I think it might be clearest to leave it as is:

- Name matching is case insensitive
- Matches are ANDed together

Given those rules, the above example could never result in a match. Does that seem reasonable?

Reply:
that sounds fine to me -- a user that put "Foo=a" and "foo=b" and expecting something is likely a rare case

Reply:
Although -- rob that would mean the implementation needs to detect this and generate a "match none" configuration?

Or we could explicitly say, having multiple entries is up to implementation to interpret -- less work on implementors, but more vague for user.

It feels like the first one is not too bad but requires a special case handling in impls.

Reply:
BTW -- we should probably add a comment that duplicate "Foo=A", "foo=B" will match nothing to make that behavior very clear, as opposed to "derived from the postulates"

Reply:
There's a lot of nuance here and it's difficult to get the wording right. For example, the following would be valid:

```
- name: foo
  type: RegularExpression
  value: [a-z]
- name: Foo
  type: RegularExpression
  value: [a-z0-9]
```

It may not actually make any sense to configure a Route that way, but a value like `z` would be valid. And similar ambiguity would likely also exist for ImplementationSpecific match types. So the only thing we can say for sure is that specifying different values with an exact match for equivalent header names would not result in a match. And once I took the time to write that out, it seemed obvious, so I just omitted it. Maybe providing the specific example you suggested would be a better example?

Reply:
Added an example here that I think helps explain how this should be interpreted.

Reply:
Actually given the regex example, we should just say if you repeat items with the same name, it will be "implementation specific" how this is implemented. Some infra may not support AND'ing regex together -- and ANDing regex is not an operator that is easy to do with a regex.

Reply:
Good point that ANDing a regex is simply not a practical expectation here. Instead of falling back to implementation specific handling of this, I think it may make the most sense simply to revert back to something resembling the original approach. Specifying that implementations must ignore subsequent matches for equivalent names and only implement the first one should be relatively easy to implement for everyone and results in a consistent experience.

Reply:
+1 makes sense, this is both easy to implement and consistent for users.